lib/fuser_mqubit.h

```diff
@@ -0,0 +1,1087 @@
+          }
+
+          n2->gate->visited = 1;
+          cur_size -= n2->qubits.size();
+          scratch.stack.pop_back();
+        }
+
+        p1->gate->visited = 1;
+        cur_size -= p1->qubits.size();
+        scratch.stack.pop_back();
+      }
+
+      n1->gate->visited = 1;
+      cur_size -= n1->qubits.size();
+      scratch.stack.pop_back();
+    }
+  }
+
+  static void GetNextAvailableGates(unsigned max_fused_size, unsigned cur_size,
+                                    const GateF& pgate1, const GateF* pgate2,
+                                    std::vector<GateA>& scratch,
+                                    std::vector<GateA*>& next_gates) {
+    next_gates.resize(0);
+
+    for (auto link : pgate1.links) {
+      if (link->next == nullptr) continue;
+
+      auto ngate = link->next->val;
+
+      if (ngate->visited > 1 || ngate->parent->unfusible) continue;
+
+      GateA next = {ngate, {}, {}};
+      next.qubits.reserve(8);
+      next.links.reserve(8);
+
+      GetAddedQubits(pgate1, pgate2, *ngate, next);
+
+      if (cur_size + next.qubits.size() > max_fused_size) continue;
+
+      scratch.push_back(std::move(next));
+      next_gates.push_back(&scratch.back());
+    }
+  }
+
+  static bool GetPrevAvailableGates(unsigned max_fused_size,
+                                    unsigned cur_size, unsigned level,
+                                    const GateF& ngate1, const GateF* ngate2,
+                                    std::vector<GateA>& scratch,
+                                    std::vector<GateA*>& prev_gates) {
+    prev_gates.resize(0);
+
+    for (auto link : ngate1.links) {
+      if (link->prev == nullptr) continue;
+
+      auto pgate = link->prev->val;
+
+      if (pgate->visited > 997 || pgate->visited == level) continue;
+
+      if (pgate->visited > 1 || pgate->parent->unfusible) {
+        prev_gates.resize(0);
+        return false;
+      }
+
+      GateA prev = {pgate, {}, {}};
+      prev.qubits.reserve(8);
+      prev.links.reserve(8);
+
+      GetAddedQubits(ngate1, ngate2, *pgate, prev);
+
+      bool all_prev_visited = true;
+
+      for (auto link : pgate->links) {
+        if (link->prev == nullptr) continue;
+
+        if (link->prev->val->visited <= 997) {
+          all_prev_visited = false;
+          break;
+        }
+      }
+
+      if (!all_prev_visited) {
+        prev_gates.resize(0);
+        return false;
+      }
+
+      if (cur_size + prev.qubits.size() > max_fused_size) continue;
+
+      if (all_prev_visited) {
+        scratch.push_back(std::move(prev));
+        prev_gates.push_back(&scratch.back());
+      }
+    }
+
+    return true;
+  }
+
+  static void GetAddedQubits(const GateF& fgate0, const GateF* fgate1,
+                             const GateF& fgate2, GateA& added) {
+    for (std::size_t i = 0; i < fgate2.qubits.size(); ++i) {
+      unsigned q2 = fgate2.qubits[i];
+
+      bool have_added_qubits = true;
+
+      for (auto q0 : fgate0.qubits) {
```

Comment:
Using `continue` would be cleaner than a flag here, I think. For example:
```
if (std::find(q2, fgate0.qubits.begin(), fgate0.qubits.end())
    == fgate0.qubits.end()) {
  continue;
}
```

Reply:
Done.